### PR TITLE
BFD-1910: Create Synthea Bene Generation Script

### PR DESCRIPTION
**Pull request by lsmitchell**
**JIRA Ticket:**
[BFD-1910](https://jira.cms.gov/browse/BFD-1910)

**User Story or Bug Summary:**

As a BFD Engineer I would like the ability for Synthea generation to be automated to support that a job can be made to generate Synthea files with one click.

---

### What Does This PR Do?

Enhances (and renames) the previous synthea validation script in order to add steps to do further validation and run synthea.

- Moves automation files into their own folder to reduce confusion with other scripts in the directory
- Renamed 'prepare-for-new-synthea-load.py' to 'prepare-and-run-synthea.py'
- Adds extensive documentation about how to run and what it does
- Added validation that all required files are readable, writable (if needed) and in the correct location
- Added function to ensure output directory is clean prior to run (or archiving old output folder if not and creating a fresh new output folder)
- Added adjustment to validation for duplicate mbi_hashes to account for the one expected dupe that existed prior to my recent synthea load
- Added function to run and log synthea national script, and fail if failure is detected in the resulting stdout

Example success run:
image

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.
* Verify file is properly pythonic

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [x] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [x] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [x] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

**Review by christrotteradhoc — COMMENTED** [review]
minor comments and a question

File: ops/ccs-ops-misc/synthetic-data/scripts/synthea-automation/prepare-and-run-synthea.py [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+        
+    bene_id_start = get_props_value(properties_file, "exporter.bfd.bene_id_start")
+    bene_id_end = int(bene_id_start) - int(number_of_benes_to_generate)
+    query = f"select count(*) from beneficiaries where bene_id <= {bene_id_start} and bene_id > {bene_id_end}"
+    result = _execute_single_count_query(db_string, query)
+    overall_validation_result = field_has_room_in_table(bene_id_start, "beneficiaries", "bene_id", result, overall_validation_result)
+        
+    clm_grp_id_start = get_props_value(properties_file, "exporter.bfd.clm_grp_id_start")
+    ## this is the end of the batch 1 relocated ids; should be nothing between the generated start and this
+    clm_grp_id_end = '-99999831003'
```
[the remark this answers is not in the document]
Comment (lsmitchell) [reply ops/ccs-ops-misc/synthetic-data/scripts/synthea-automation/prepare-and-run-synthea.py:213]: No, this is the set of old clm_grp_ids that got shoved into the 'upper bounds' so they were out of the way 'forever'. Our clm_grp_id ends at -1617744710 and increments from there for all new generations, so we should never have to worry about hitting -99999831003 where the reorganized old data got put.

File: ops/ccs-ops-misc/synthetic-data/scripts/synthea-automation/prepare-and-run-synthea.py [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+# beneficiary files. 
+#
+# Args:
+# 1: previous end state properties file location
+# 2: number of beneficiaries to be generated
+# 3: file system location of synthea folder
+# 4: db string for target environment DB, in this format: postgres://<dbName>:<db-pass>@<aws db url>:5432/fhirdb
+# 5: (optional) boolean to skip validation if True, useful if re-generating a bad batch, True or False
+#
+# Example runstring (with test db stored as an env variable TEST_DB): python3 prepare-and-run-synthea.py ~/end-state.properties ~/Git/synthea/ 2000 $TEST_DB
```
[the remark this answers is not in the document]
Comment (lsmitchell) [reply ops/ccs-ops-misc/synthetic-data/scripts/synthea-automation/prepare-and-run-synthea.py:12]: Good catch! Fixed order and updated the doc refs that pointed to 2 and 3 to be correct in the below parts.

File: ops/ccs-ops-misc/synthetic-data/scripts/synthea-automation/prepare-and-run-synthea.py [hunk 3]
```diff
@@ -0,0 +1,411 @@
+#
+# Script for preparing a synthea repo for generating a number of
+# beneficiary files. 
+#
+# Args:
+# 1: previous end state properties file location
+# 2: number of beneficiaries to be generated
+# 3: file system location of synthea folder
+# 4: db string for target environment DB, in this format: postgres://<dbName>:<db-pass>@<aws db url>:5432/fhirdb
+# 5: (optional) boolean to skip validation if True, useful if re-generating a bad batch, True or False
```
[the remark this answers is not in the document]
Comment (lsmitchell) [reply ops/ccs-ops-misc/synthetic-data/scripts/synthea-automation/prepare-and-run-synthea.py]: Added!

**Review by christrotteradhoc — APPROVED** [review]
Good stuff!

**Review by mjburling — APPROVED** [review]

**Pull request merged**